text/1004-transactions-api.rst

```diff
@@ -0,0 +1,619 @@
+2. Repeating transactions would make less incentive to keep transactions
+   open for a long time (e.g. while accessing slow network resources like
+   external API), which is a problem of itself.
+3. Even if we never have failed concurrent updates we would want
+   seamless reconnect on connection failures (i.e. server restart,
+   primary/replica change, etc.)
+4. To make learning curves shorter I think we should intentionally
+   inject failures. This is needed so that users quickly find out that side
+   effects of their transactions are in effect several times.
+
+So while increasing learning curve, we fix heisenbugs and simplify
+operations.
+
+Failure Injection
+-----------------
+
+The following is proposed to be done by default:
+
+Collect statitics of how many queries are executed in the previous
+second and on each new request trigger a failure with the probability of
+``1/n`` where ``n`` is the number of requests in the previous second. We
+still need to figure out whether ``n`` counts queries, transactions, or
+mutable queries/transactions (and have a list of exceptions, perhaps:
+dump+restore+migrations).
+
+The idea is that there will be ~1 failure per second. So on local
+instance when testing manually it would hit almost every request (which
+is fine as repeating them shouldn't be prohibitively costly). But under
+a huge load of thousands requests per second, one retry per second
+doesn't influence anything so even for production and/or benchmarks this
+is fine.
+
+I think it should be disabled by an explicit command-line argument like
+``--disable-failure-injection`` but might be tweaked with configuration
+settings?
+
+
+Alternatives
+============
+
+Alternative Names
+-----------------
+
+For ``retry`` method:
+
+1. ``db.atomic(t => t.execute(..))``
+2. ``db.mutate(transaction => transaction.execute(..))``
+3. ``db.apply(transaction => transaction.execute(..))``
+4. ``db.unit_of_work(t => t.execute(..))``
+5. ``db.block(t => t.execute(..))``
+6. ``db.try(transaction => transaction.execute(..))``
+7. ``db.retry_transaction(t => t.execute(..))``
+
+For ``try_transaction`` method:
+
+1. ``with db.raw_transaction() as t:``
+2. ``with db.plain_transaction() as t:``
+3. ``with db.unreliable_transaction() as t:``
+4. ``with db.single_transaction() as t:``
+
+
+Alternative Python API
+----------------------
+
+For python API we could support funcional API:
+
+.. code-block:: python
+
+    def handler(req, db: edgedb.Pool):
+        await db.retry(my_tx, req)
+
+    async def my_tx(transaction, req):
+        let val = await transaction.fetch("...")
+        await transaction.execute("...", process_value(val))
+
+And/or decorator API:
+
+.. code-block:: python
+
+    def handler(req, db: edgedb.Pool):
+        do_something(req)
+
+        @db.retry()
+        def my_tx(transaction):
+            let val = transaction.fetch("...")
+            transaction.execute("...", process_value(val))
```

Comment:
Since we do want sync and async APIs look similar, I suggest adding an explicit `my_tx()` call.

Reply:
If we're adding a call, then it's a different API, I think. Because then it's tempting to move function out of the closure, and ... Anyway this is alternative API to what we're implementing, so I don't think it is important.